lib/testutils/src/lib.rs

```diff
@@ -340,6 +339,88 @@ impl TestWorkspace {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DirEntry {
+    File { contents: Vec<u8>, executable: bool },
+    Symlink { target: String },
+    GitSubmodule { commit: CommitId },
+}
+
+impl DirEntry {
+    pub fn empty_file() -> Self {
+        DirEntry::File {
+            contents: vec![],
+            executable: false,
+        }
+    }
+
+    pub fn text_file(contents: &str) -> Self {
+        DirEntry::File {
+            contents: contents.as_bytes().to_vec(),
+            executable: false,
+        }
+    }
+
+    pub fn binary_file(contents: Vec<u8>) -> Self {
+        DirEntry::File {
+            contents,
+            executable: false,
+        }
+    }
+
+    pub fn symlink(target: &str) -> Self {
+        DirEntry::Symlink {
+            target: target.to_owned(),
+        }
+    }
+
+    pub fn submodule(commit: CommitId) -> Self {
+        DirEntry::GitSubmodule { commit }
+    }
+
+    pub fn executable(self, executable: bool) -> Self {
+        match self {
+            Self::File { contents, .. } => Self::File {
+                contents,
+                executable,
+            },
+            DirEntry::Symlink { .. } => panic!("symlinks cannot be executable"),
+            DirEntry::GitSubmodule { .. } => panic!("submodules cannot be executable"),
+        }
+    }
+
+    pub fn is_executable(&self) -> bool {
+        match self {
+            DirEntry::File { executable, .. } => *executable,
+            DirEntry::Symlink { .. } => panic!("symlinks cannot be executable"),
+            DirEntry::GitSubmodule { .. } => panic!("submodules cannot be executable"),
+        }
+    }
+
+    pub fn text_contents(&self) -> Option<&str> {
```

Comment:
Same here: can the callers use `match` instead?

Reply:
Not really because `contents: Vec<u8>` does not match nicely against strings. But this function was not used and perhaps it won't ever be needed because `DirEntry` is so far only consumed by `create_tree()`. I've removed the function.